tsdb/chunks/chunks.go

```diff
@@ -252,63 +254,131 @@ func MergeOverlappingChunks(chks []Meta) ([]Meta, error) {
 			last++
 			continue
 		}
-		nc := &newChks[last]
-		if c.MaxTime > nc.MaxTime {
-			nc.MaxTime = c.MaxTime
-		}
-		chk, err := MergeChunks(nc.Chunk, c.Chunk)
+		dividedChunkOne, dividedChunkTwo, err := MergeChunks(newChks[last].Chunk, c.Chunk, MaxSamplesInChunk)
 		if err != nil {
 			return nil, err
 		}
-		nc.Chunk = chk
+		newChks[last] = *dividedChunkOne
+		if dividedChunkTwo != nil {
+			newChks = append(newChks, *dividedChunkTwo)
+			last++
+		}
 	}
-
 	return newChks, nil
 }
 
 // MergeChunks vertically merges a and b, i.e., if there is any sample
 // with same timestamp in both a and b, the sample in a is discarded.
-func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
+// Also breaks chunks that are larger than a particular limit.
+func MergeChunks(a, b chunkenc.Chunk, limit int) (*Meta, *Meta, error) {
+	newMetaArray := make([]Meta, 0)
 	newChunk := chunkenc.NewXORChunk()
 	app, err := newChunk.Appender()
+	appPtr := &app
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	ait := a.Iterator(nil)
 	bit := b.Iterator(nil)
 	aok, bok := ait.Next(), bit.Next()
+	numSamples := 0
+	appendAndReset := func(newMetaArrayPtr *[]Meta, newChunk *chunkenc.XORChunk, mint int64, maxt int64) (*chunkenc.XORChunk, *chunkenc.Appender, error) {
```

Comment:
this is way to complex. Also why not making it separate function then?

Reply:
because you had told me to use anonymous functions [before](https://github.com/prometheus/prometheus/pull/6760#discussion_r376163601) too.
P.S. Thanks for the review!

Reply:
Well, don't take everything so explicit. Try to have your own common sense here (: If something is readable or not (: